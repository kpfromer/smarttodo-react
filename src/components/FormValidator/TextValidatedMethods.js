import validateIsEmail from "validator/lib/isEmail";

export const required = isRequired => ({
  hint: name => isRequired ? `${name} is required` : '',
  validate: value => isRequired ? !!value : true
});

export const minLength = minValue => ({
  hint: name => `${name} must be ${minValue + 1} characters or more`,
  validate: value => value.length > minValue
});

export const maxLength = maxValue => ({
  hint: name => `${name} must be ${maxValue - 1} characters or less`,
  validate: value => value.length < maxValue
});

export const email = isEmail => ({
  hint: () => isEmail ? `Must be a valid email` : '',
  validate: value => isEmail ? validateIsEmail(value) : true
});