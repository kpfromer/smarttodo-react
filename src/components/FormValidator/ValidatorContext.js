import { createContext } from "react";

export const ValidatorContext = createContext({
  initializeInput: () => {},
  removeInput: () => {},
  updateFormInput: () => {}
});