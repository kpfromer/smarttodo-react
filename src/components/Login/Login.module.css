.container {
  flex-grow: 1;
  display: flex;
}

.card {
  margin: auto;
  max-width: 345px;
}

.cardActions {
  display: flex;
  justify-content: flex-end;
}
