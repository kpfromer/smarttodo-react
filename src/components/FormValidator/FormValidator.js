import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { ValidatorContext } from './ValidatorContext';

// TODO: refactor
// TODO: error on touch (when user enters then immediately leaves)
// TODO: validate all elements on submit, for example required empty inputs would become red
export default class FormValidator extends Component {

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    validateOnSubmit: PropTypes.bool,
    onError: PropTypes.func
  };

  static defaultProps = {
    validateOnSubmit: false,
    onError: () => {
    }
  };

  updateFormInput = ({ name, value, error }) => {
    this.setState(prevState => ( {
      initializedInputs: prevState.initializedInputs.map(input => {
        if (input.name === name) {
          return {
            ...input,
            value,
            edited: true
          };
        }
        return input;
      })
    } ));

    if (error) {
      this.setState(prevState => ( {
        errors: [
          ...prevState.errors,
          name
        ]
      } ));
    } else {
      this.setState(prevState => ( {
        errors: prevState.errors.filter(erroredInputName => erroredInputName !== name)
      } ));
    }
  };

  initializeInput = (name, value, triggerValidation) =>
    this.setState(prevState => ( {
      initializedInputs: [...prevState.initializedInputs, { name, value, triggerValidation, edited: false }]
    } ));

  removeInput = name =>
    this.setState(prevState => ( {
      errors: prevState.errors.filter(errorName => errorName !== name),
      initializedInputs: prevState.initializedInputs.filter(input => input.name !== name)
    } ));

  state = {
    context: {
      initializeInput: this.initializeInput,
      removeInput: this.removeInput,
      updateFormInput: this.updateFormInput,
    },
    initializedInputs: [],
    errors: []
  };

  handleSubmit = event => {
    event.preventDefault();

    const { errors, initializedInputs } = this.state;

    if (this.props.validateOnSubmit) {
      initializedInputs.forEach(input => input.triggerValidation(input.value));
    }

    if (errors.length === 0 && initializedInputs.every(input => input.edited)) {
      this.props.onSubmit();
    } else {
      this.props.onError();
    }
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <ValidatorContext.Provider value={this.state.context}>
          {this.props.children}
        </ValidatorContext.Provider>
      </form>
    );
  }
}
