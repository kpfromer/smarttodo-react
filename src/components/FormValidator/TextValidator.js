import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { isArray, isPlainObject } from 'lodash';

import { ValidatorContext } from './ValidatorContext';
import TextField from '@material-ui/core/TextField';

const validateShape = PropTypes.shape({
  text: PropTypes.string.isRequired,
  validate: PropTypes.func.isRequired
}).isRequired;

export class TextValidator extends Component {

  static propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    validate: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(validateShape),
      validateShape
    ]).isRequired,
    validateOnBlur: PropTypes.bool,
    textFieldProps: PropTypes.object,
    name: PropTypes.string,
    updateFormInput: PropTypes.func,
    initializeInput: PropTypes.func,
    removeInput: PropTypes.func,
  };

  static defaultProps = {
    validateOnBlur: false
  };

  state = {
    error: false,
    errorText: ''
  };

  validateValue = value => {
    const { updateFormInput, validate, name } = this.props;

    let error = false;
    let errorText = '';

    const validatationObject = validateObj => {
      const isError = !validateObj.validate(value);
      if (isError) {
        error = true;
        errorText = validateObj.text;
      }
    };

    if (isArray(validate)) {
      for (const validateObj of validate) {
        validatationObject(validateObj);
        if (error) {
          break;
        }
      }
    } else if (isPlainObject(validate)) {
      validatationObject(validate);
    } else {
      error = !validate(value);
    }

    if (updateFormInput) {
      updateFormInput({
        name,
        value,
        error
      });
    }
    this.setState({ error, errorText });

    return { error };
  };

  componentDidMount() {
    const { initializeInput, name, value } = this.props;
    if (initializeInput) {
      initializeInput(name, value, this.validateValue);
    }
  }

  componentWillUnmount() {
    const { removeInput, name } = this.props;
    if (removeInput) {
      removeInput(name);
    }
  }

  componentDidUpdate(prevProps) {
    const { value: oldValue, name: oldName } = prevProps;
    const { value: newValue, name: newName, removeInput, initializeInput } = this.props;

    if (oldValue !== newValue) {
      this.validateValue(newValue);
    }

    if (oldName !== newName) {
      removeInput(oldName);
      initializeInput(newName);
    }
  }

  handleChange = event => {
    const { value } = event.target;
    const { error } = this.validateValue(value);
    this.props.onChange({ value, error });
  };

  handleBlur = event => {
    if (this.props.validateOnBlur) {
      this.validateValue(event.target.value);
    }
  };

  render() {
    const { value, textFieldProps } = this.props;
    const { error, errorText } = this.state;
    return (
      <TextField
        {...textFieldProps}
        onBlur={this.handleBlur}
        value={value}
        error={error}
        onChange={this.handleChange}
        helperText={errorText}
      />
    );
  }
}

export default props =>
  <ValidatorContext.Consumer>
    {
      ({ updateFormInput, initializeInput, removeInput }) =>
        <TextValidator
          {...props}
          updateFormInput={updateFormInput}
          initializeInput={initializeInput}
          removeInput={removeInput}
        />
    }
  </ValidatorContext.Consumer>
