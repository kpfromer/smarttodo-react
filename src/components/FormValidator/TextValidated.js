import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { email, maxLength, minLength, required } from './TextValidatedMethods';
import TextValidator from "./TextValidator";

export default class TextValidated extends Component {

  static propTypes = {
    hintName: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    name: PropTypes.string,
    textValidatorProps: PropTypes.object,
    required: PropTypes.bool,
    minLength: PropTypes.number,
    maxLength: PropTypes.number,
    email: PropTypes.bool
  };

  static defaultProps = {
    required: null,
    minLength: null,
    maxLength: null,
    email: null
  };

  validatedMethods = {
    required,
    minLength,
    maxLength,
    email
  };

  render() {
    const { name, value, onChange, hintName, textValidatorProps, ...validationTypes } = this.props;
    const validate = Object.keys(validationTypes)
      .filter(validationName => this.props[validationName] !== null)
      .map(validationName => {
        if (!Object.keys(this.validatedMethods).includes(validationName)) {
          throw new Error(`Invalid property of TextValidated: ${validationName}`);
        }

        const validateValue = this.props[validationName];

        if (!validateValue) {
          throw new Error(`Invalid property of TextValidated: ${validationName}`);
        }

        const { hint, validate } = this.validatedMethods[validationName](validateValue);
        return {
          text: hint(hintName),
          validate
      }
    });

    return (
      <TextValidator
        {...textValidatorProps}
        name={name}
        value={value}
        onChange={onChange}
        validate={validate}
      />
    );
  }
}
