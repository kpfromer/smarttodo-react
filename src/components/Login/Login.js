import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Card from '@material-ui/core/Card';
import { tryLogin } from "../../actions/login";
import { Redirect, withRouter } from "react-router-dom";
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import FormValidator from "../FormValidator/FormValidator";

import styles from './Login.module.css';
import TextValidated from "../FormValidator/TextValidated";
import { changeSnackbar } from "../../actions/snackbar";

export class Login extends PureComponent {

  static propTypes = {
    login: PropTypes.func.isRequired,
    changeSnackbar: PropTypes.func.isRequired
  };

  state = {
    username: '',
    password: '',
    redirectToReferrer: false
  };

  loginUser = (username, password) =>
    this.props.login(username, password)
      .then(() => {
        this.props.changeSnackbar({
          message: 'Successfully logged in. Redirecting!',
          variant: 'success'
        });
        this.setState({ redirectToReferrer: true });
      })
      .catch(response => {
        this.props.changeSnackbar({
          message: response.error,
          variant: 'error'
        });
        this.setState({ password: '' })
      });

  handleSubmit = () => {
    const { username, password } = this.state;
    this.loginUser(username, password);
  };

  updateValue = valueName => event => {
    const { value } = event;
    this.setState({
      [valueName]: value
    })
  };

  validate = name => ([
    {
      validate: value => !!value,
      text: `${name} is required`
    },
    {
      validate: value => value.length > 3,
      text: `${name} must be more than 3 characters`
    }
  ]);

  render() {
    if (this.state.redirectToReferrer) {
      // TODO: add flash message to and from
      // https://github.com/mjrussell/redux-auth-wrapper/issues/64
      return <Redirect to="/todo"/>;
    }

    return (
      <div className={styles.container}>
        <Card className={styles.card}>
          <FormValidator onSubmit={this.handleSubmit}>
            <CardContent>
              <Typography variant="title">
                Login
              </Typography>
              <TextValidated
                name="username"
                hintName="Username"
                required
                minLength={3}
                value={this.state.username}
                onChange={this.updateValue('username')}
                textFieldProps={{
                  autoFocus: true,
                  margin: 'dense',
                  label: 'Username',
                  type: 'text',
                  // autoComplete: 'username',
                  fullWidth: true
                }}
              />
              <TextValidated
                name="password"
                hintName="Password"
                required
                minLength={3}
                value={this.state.password}
                onChange={this.updateValue('password')}
                textFieldProps={{
                  autoFocus: true,
                  margin: 'dense',
                  label: 'Password',
                  type: 'password',
                  // autoComplete: 'current-password',
                  fullWidth: true
                }}
              />
            </CardContent>
            <CardActions className={styles.cardActions}>
              <Button type="submit" size="medium" color="primary">
                Submit
              </Button>
            </CardActions>
          </FormValidator>
        </Card>
      </div>
    );
  }
}

export default connect(null, {
  login: tryLogin,
  changeSnackbar
})(withRouter(Login));