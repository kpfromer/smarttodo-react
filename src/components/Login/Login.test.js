import { Login } from "./Login";
import { SetupComponent } from "react-component-setup";
import TextValidator from "../FormValidator/TextValidator";
import FormValidator from "../FormValidator/FormValidator";
import Snackbar from "@material-ui/core/Snackbar";

const { shallow: setup } = SetupComponent({
  component: Login
});

describe('Login', () => {
  let wrapper, mockLogin, get;

  beforeEach(() => {
    get = name => wrapper.find(TextValidator).findWhere(node => node.prop('name') === name);

    mockLogin = jest.fn();
    ({ wrapper } = setup({
      login: mockLogin
    }));
  });

  [
    {
      type: 'username',
      setup: () => 'example username'
    },
    {
      type: 'password',
      setup: () => 'example password'
    }
  ].forEach(value => {
    const { type, setup } = value;

    describe(`user enters ${type}`, () => {
      let input, newValue;

      beforeEach(() => {
        input = get(type);

        newValue = setup();

        input.props().onChange({ value: newValue });
        wrapper.update();
        input = get(type);
      });

      it(`${type} input value changes`, () => {
        expect(input).toHaveProp('value', newValue);
      });
      it('validates input', () => {
        expect(input.props().validate).toMatchSnapshot();

        const isRequired = input.props().validate.find(value => value.text.includes('is required')).validate;
        const moreThanThree = input.props().validate.find(value => value.text.includes('must be more than 3 characters')).validate;

        expect(isRequired('hello')).toBe(true);
        expect(isRequired('')).toBe(false);
        expect(moreThanThree('3333')).toBe(true);
        expect(moreThanThree('fff')).toBe(false);
      });
    });
  });

  describe('user submits login form with username and password filled', () => {
    let username, password, findTextValidatorByName;

    beforeEach(() => {
      findTextValidatorByName = name => wrapper.find(TextValidator).findWhere(node => node.prop('name') === name);

      username = 'exampleUsername';
      password = 'examplePassword';

      findTextValidatorByName('username').props().onChange({ value: username });
      findTextValidatorByName('password').props().onChange({ value: password });
      wrapper.update();
    });

    describe('if credentials are valid', () => {
      beforeEach(async () => {
        const promise = Promise.resolve();
        mockLogin.mockReturnValue(promise);
        wrapper.find(FormValidator).props().onSubmit();
        await promise;
        wrapper.update();
      });

      it('logins in user', () => {
        expect(mockLogin).toHaveBeenCalledWith(username, password);
      });
      it('redirects user to /todo', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('if credentials are invalid', () => {
      let error;

      beforeEach(() => {
        error = 'Username/password were not found!';
        const promise = Promise.reject({
          error
        });
        mockLogin.mockReturnValue(promise);
        wrapper.find(FormValidator).props().onSubmit();
        return promise
          .then(() => {}) // See https://github.com/airbnb/enzyme/issues/450#issuecomment-341244926
          .catch(() => wrapper.update());
      });

      it('removes password', () => {
        expect(findTextValidatorByName('password')).toHaveProp('value', '');
      });
      it('displays a snackbar message with the api error', () => {
        expect(wrapper.find(Snackbar)).toMatchSnapshot();
      });
      describe('error snackbar', () => {
        it('closes when the X icon is clicked', () => {
          wrapper.find(Snackbar).props().onClose();
          wrapper.update();
          expect(wrapper.find(Snackbar)).toHaveProp('open', false);
        });
      });
    });
  });
});