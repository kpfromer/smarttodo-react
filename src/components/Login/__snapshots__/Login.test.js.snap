// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login user enters password validates input 1`] = `
Array [
  Object {
    "text": "Password is required",
    "validate": [Function],
  },
  Object {
    "text": "Password must be more than 3 characters",
    "validate": [Function],
  },
]
`;

exports[`Login user enters username validates input 1`] = `
Array [
  Object {
    "text": "Username is required",
    "validate": [Function],
  },
  Object {
    "text": "Username must be more than 3 characters",
    "validate": [Function],
  },
]
`;

exports[`Login user submits login form with username and password filled if credentials are invalid displays a snackbar message with the api error 1`] = `
<WithStyles(Snackbar)
  anchorOrigin={
    Object {
      "horizontal": "center",
      "vertical": "bottom",
    }
  }
  autoHideDuration={3000}
  onClose={[Function]}
  open={true}
>
  <WithStyles(SpecialSnackbar)
    message="Username/password were not found!"
    onClose={[Function]}
    variant="error"
  />
</WithStyles(Snackbar)>
`;

exports[`Login user submits login form with username and password filled if credentials are valid redirects user to /todo 1`] = `
<Redirect
  push={false}
  to="/todo"
/>
`;
